// ng-toggle-button/esm2020/lib/ng-toggle.module.mjs
import { NgModule } from '@angular/core';
import { NgToggleComponent } from './ng-toggle.component';
import { CommonModule } from '@angular/common';
import { NgToggleConfig } from './ng-toggle.config';
import * as i0 from "@angular/core";
export class NgToggleModule {
    static forRoot(config = {}) {
        return {
            ngModule: NgToggleModule,
            providers: [
                {
                    provide: NgToggleConfig,
                    useValue: config
                }
            ]
        };
    }
}
NgToggleModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.8", ngImport: i0, type: NgToggleModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgToggleModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.8", ngImport: i0, type: NgToggleModule, declarations: [NgToggleComponent], imports: [CommonModule], exports: [NgToggleComponent] });
NgToggleModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.8", ngImport: i0, type: NgToggleModule, providers: [NgToggleConfig], imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.8", ngImport: i0, type: NgToggleModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgToggleComponent],
                    imports: [
                        CommonModule
                    ],
                    exports: [NgToggleComponent],
                    providers: [NgToggleConfig]
                }]
        }] });
